psutil/tests/__init__.py

```diff
@@ -260,6 +262,27 @@ def attempt(exe):
 
 
 PYTHON_EXE = _get_py_exe()
+
+
+def _get_base_py_exe():
+    # On windows, starting with python 3.7,
+    # virtual environments use a venvlauncher startup process
+    # This does not play well when counting spawned processes or
+    # when relying on the pid of the spawned process to do some checks
+    # e.g. connection check per pid
+    # let's use the base python in this case.
+    base = getattr(sys, "_base_executable", None)
+    needs_base = \
+        WINDOWS and \
+        sys.version_info >= (3, 7) and \
+        IS_VIRTUALENV and \
+        base is not None
+    if needs_base:
+        return base
+    return PYTHON_EXE
+
+
+PYTHON_BASE_EXE = _get_base_py_exe()
```

Comment:
Unless there are circumstances in which it is necessary to use one instead of the other, I think it's simpler for the caller to simply think in terms of a single "PYTHON_EXE", without introducing a new public constant. Why not integrate this logic into `_get_py_exe()` instead? Something like (not tested):

```diff
+++ b/psutil/tests/__init__.py
@@ -196,6 +196,21 @@ SKIP_SYSCONS = (MACOS or AIX) and os.getuid() != 0
 # --- misc
 
 
+def _get_win_venv_py_exe():
+    # On windows, starting with python 3.7, virtual environments use a
+    # venvlauncher startup process This does not play well when counting
+    # spawned processes or when relying on the pid of the spawned
+    # process to do some checks e.g. connection check per pid let's use
+    # the base python in this case.
+    if (
+        WINDOWS and
+        sys.version_info >= (3, 7) and
+        IS_VIRTUALENV and
+        getattr(sys, "_base_executable", None)
+    ):
+        return sys._base_executable
+
+
 def _get_py_exe():
     def attempt(exe):
         try:
@@ -206,7 +221,9 @@ def _get_py_exe():
         else:
             return exe
 
-    if GITHUB_ACTIONS:
+    if WINDOWS and _get_win_venv_py_exe():
+        return _get_win_venv_py_exe()
+    elif GITHUB_ACTIONS:
         if PYPY:
             return which("pypy3") if PY3 else which("pypy")
         elif FREEBSD:
```

If possible, I would prefer a unified solution, mostly because I think there currently isn't a specific build for Windows that uses a virtualenv, so I'm afraid we may accidentally break tests again in the future.

Reply:
> Unless there are circumstances in which it is necessary to use one instead of the other, I think it's simpler for the caller to simply think in terms of a single "PYTHON_EXE", without introducing a new public constant

Agreed, however, using the base executable might require setting a specific environment variable to get things working properly (c.f. `__PYVENV_LAUNCHER__ ` in `pyrun`).
I'll check if I can get things unified.

> I think there currently isn't a specific build for Windows that uses a virtualenv, so I'm afraid we may accidentally break tests again in the future.

All windows tests are moved to cibuildwheel which always uses a virtual environment. This simplifies CI setup a bit and would allow to get a dev workflow using `nox` or `tox`, or just a virtual environment easily.

Reply:
> All windows tests are moved to cibuildwheel which always uses a virtual environment.

I'm curious: was there a specific test failure that led to this PR?

Reply:
Wow! That's a lot! Glad you're fixing it.

Reply:
I reworked the PR to simplify all of this.
The only test where I expect things to fail is test_misc.TestScripts when ran from a virtual environment but those tests are just skipped in CI.
Either I try to get them running in CI here or leave it as-is for now ?

Reply:
I opened #2211 for `test_misc.TestScripts`, maybe good to get it merged as a pre-requisite for this PR ?